### ISSUE TITLE
Improved text selection

### ISSUE THREAD

**Issue by jonahwilliams** [flutter/flutter#81839]
Currently the flutter text selection is fairly limited. Historically this is based on the same code that enables text interaction in text edit environments. This has lead to two peculiarities: 

* It is text field specific, instead of being global. This prevents copying of text together that happens to be grouped into different `SelectableText` widgets.

* It is opt-in. This makes sense for mobile where touch gestures are heavily overloaded, but on desktop and desktop web applications drag gestures universally perform text selection. In general, requiring an opt-in is fairly user hostile.

These two problems can be addressed separately,  but together would make text selection behave as a user would expect.

In order to make text selection global, we'd need to hoist the selection logic into a new widget, lets call it a `SelectionArea`. This could be introduced by default under the `WidgetsApp`. This widget would introduce the correct drag gesture detector which wrapped the entire screen. It would also provide itself as an inherited widget, allowing text and other widgets types to register their text contents and screen location with. When a drag gesture is performed, it could determine whether the drag "crossed" (from a document perspective) the text selection and highlight it using a single overlay.

Some open questions on this:

* Can we make the screen sized drag gesture detector always lose to more specific gesture detectors? How do we avoid accidentally competing with other contents.
* How do we handle selection through an infinite scrollable / slivers? It seems like screen coordinates would not be sufficient, we'd also need to know some relative coordinates.
* How do we handle non-text data types? Selection could be made generic to support images or other compatible clipboard types.
* How do users opt-out of selection?

Once the selection logic is hoisted into the new widget, the `SelectableText` widget can be hollowed out a bit. The registration code can be moved into the text RenderObject, with the Text/SelectableText widgets responsible for retrieving the inherited widget. A single field on Text could configure this selection behavior, with it enabled by default on desktop and web and disabled by default elsewhere. SelectableText would be updated to always set this flag to true.

design doc: http://flutter.dev/go/global-selection

sub task:

- [ ] Enable supported devices for TapGestureRecognizer https://github.com/flutter/flutter/issues/96556
- [ ] Modularize auto scroll logic in Reorderable list https://github.com/flutter/flutter/issues/96562

**Comment by HansMuller** [comment]
CC @chunhtai @LongCatIsLooong @justinmc 

We have been talking about this for a while and agree that it's a good idea.  It's a big project that addresses a common complaint about Flutter apps on desktop and particularly on the web.

**Comment by jonahwilliams** [comment]
From some further investigation:

* We need to figure out how this interacts with modals/overlays/routes. My guess is pushing/popping new routes should auto clear the selection.

* We need a new text overlay. The current one is based on having a TextEditingValue. could be done in a single overlay at the widgets app level, or pushed to each individual widget.

**Comment by jonahwilliams** [comment]
Some answers through the PR:

- Introduce the selection area at the route level, this prevents selection from accidentally working cross page.
- Have the Paragraph objects responsible for drawing the selection highlights themselves
- Perhaps have the paragraph objects handle computing the selection range, to make it easier to account for nest LTR/RTL

**Comment by chunhtai** [comment]
note to myself

example code for testing

<details>
<summary> code </summary>

```
// Copyright 2019 The Flutter team. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';
void main() => runApp(MaterialApp(debugShowCheckedModeBanner: false,home:AppBarDemo()));
class AppBarDemo extends StatelessWidget {
  const AppBarDemo({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: IconButton(
          tooltip: MaterialLocalizations.of(context).openAppDrawerTooltip,
          icon: const Icon(Icons.menu),
          onPressed: () {},
        ),
        title: Text(
          'my title'
        ),
        actions: [
          IconButton(
            tooltip: 'icon 1',
            icon: const Icon(
              Icons.favorite,
            ),
            onPressed: () {},
          ),
          IconButton(
            tooltip: 'icon 2',
            icon: const Icon(
              Icons.search,
            ),
            onPressed: () {},
          ),
          PopupMenuButton<Text>(
            itemBuilder: (context) {
              return [
                PopupMenuItem(
                  child: Text(
                      'first'
                  ),
                ),
                PopupMenuItem(
                  child: Text(
                    'second'
                  ),
                ),
                PopupMenuItem(
                  child: Text(
                    'third'
                  ),
                ),
              ];
            },
          )
        ],
      ),
      body: Center(
        child: Text(
          'hometab'
        ),
      ),
    );
  }
}
```
</details>

**Comment by chunhtai** [comment]
blocked by https://github.com/flutter/flutter/issues/95974
in order to draw the selection handle, the layer link need to jump from one render object to another render object within the same frame, and that causes a crash due to the linked issue

**Comment by chunhtai** [comment]
A note to myself:

macOS: one can store NSAttributeString into UIPasteboard for rich text,  or use named clipboard for custom data

windows: one can use CF_RTF with SetClipboardData for rich text. or different format type

It looks like none of platform support text with media unless you use custom data type and interpret yourself. It looks like copying TextSpan structure into rich text format in different platform may be a good first step.

**Comment by chunhtai** [comment]
For Linux GTK there doesn't seem to be an API to write rich text into the clipboard, but there is a method to requrest rich text back `gtk_clipboard_request_rich_text`

@robert_ancell do you maybe have some insights on this? Is there a way to store rich text into clipboard?

**Comment by robert-ancell** [comment]
> @robert-ancell do you maybe have some insights on this? Is there a way to store rich text into clipboard?

I think you do something like this:

```c
void data_cb((GtkClipboard *clipboard, GtkSelectionData *selection_data, guint info, gpointer data)
{
   if (info == 1) {
     // return a string
   } else if (info == 2) {
     // return a rich text string
   }
}

GtkTargetEntry entries[] = { { .target = "UTF8_STRING", .info = 1 }, {.target = "text/rtf", .info = 2} };
gtk_clipboard_set_with_data(clipboard, entries, 2, data_cb, NULL, NULL);
```
`gtk_clipboard_request_rich_text` just automatically handles the formats that `GtkTextBuffer` supports.

**Comment by tgucio** [comment]
I guess a good question is whether one wants to go the rich text road at all. Nowadays most applications appear to correctly support HTML clipboard. Any RenderObject can be turned to image data with .toImage(), then it's just a matter of encoding as PNG and converting to an <img /> element with inline data: <img src="data:image/png;base64,..." />

Setting that on each platform can look more or less like:

<details>
  <summary>Android</summary>
  
```
      kSetDataMethod -> {
        val text = call.argument("text") as String?
        val html = call.argument("html") as String?
        if (text != null) {
          if (html != null) {
            var clip: ClipData = ClipData.newHtmlText("Flutter app content", text, html)
            clipboard.setPrimaryClip(clip)
          } else {
            var clip: ClipData = ClipData.newPlainText("Flutter app content", text)
            clipboard.setPrimaryClip(clip)
          }
          result.success(null);
        } else {
          result.error("set_data", "Missing text", null)
        }
      }
```
</details>

<details>
  <summary>iOS</summary>
  
```
      case kSetDataMethod:
        let text = arguments?["text"] as? String
        let html = arguments?["html"] as? String
        var items: [[String : String]] = []
        if text != nil {
          items.append([kUTTypePlainText as String: text!])
        }
        if html != nil {
          items.append([kUTTypeHTML as String: html!])
        }
        pasteboard.items = items
        result(true)
        break
```
</details>

<details>
  <summary>macOS</summary>
  
```
        let text = arguments?["text"] as? String
        let html = arguments?["html"] as? String

        var types: [NSPasteboard.PasteboardType] = []
        if text != nil {
          types.append(NSPasteboard.PasteboardType.string)
        }
        if html != nil {
          types.append(NSPasteboard.PasteboardType.html)
        }
        if types.isEmpty {
          result(false)
          break
        }

        pasteboard.clearContents()
        pasteboard.declareTypes(types, owner: nil)
        if text != nil {
          pasteboard.setString(text!, forType: .string)
        }
        if html != nil {
          pasteboard.setString(html!, forType: .html)
        }
        result(true)
        break
```
</details>

Windows (Win32) needs two tricks:
* registering a clipboard format called "HTML Format"
* adding horrendous headers that require calculating offsets: https://docs.microsoft.com/en-us/windows/win32/dataxchg/html-clipboard-format

<details>
  <summary>Windows</summary>
  
```
  if (!::SetClipboardData(CF_UNICODETEXT, locked_memory_text.get())) {
    return false;
  }

  UINT CF_HTML = ::RegisterClipboardFormatA("HTML Format");
  if (CF_HTML == 0) {
    return false;
  }
  if (!::SetClipboardData(CF_HTML, locked_memory_html.get())) {
    return false;
  }
```
</details>

Linux is a bit more tricky as GTK needs a callback to supply data other than plain text:
<details>
  <summary>Linux</summary>
  
```
struct ClipboardContents {
  gchar* textHtml = nullptr;
  gchar* textPlain = nullptr;
};

static GdkAtom atomTextHtml = gdk_atom_intern_static_string(kTextHtml);
static GdkAtom atomUtf8String = gdk_atom_intern_static_string("UTF8_STRING");
static GdkAtom atomTextPlain = gdk_atom_intern_static_string("text/plain");
static GdkAtom atomTextPlainUtf8 = gdk_atom_intern_static_string("text/plain;charset=utf-8");

static void
clipboard_get_cb(GtkClipboard *clipboard,
                 GtkSelectionData *selection_data,
                 guint info,
                 gpointer data) {
  ClipboardContents* contents = (ClipboardContents *) data;

  GdkAtom target = gtk_selection_data_get_target(selection_data);
  if (target == atomTextHtml) {
    if (contents->textHtml != nullptr) {
      gtk_selection_data_set(selection_data, GDK_SELECTION_TYPE_STRING,
                            8, (unsigned char*) contents->textHtml, strlen(contents->textHtml));
    } else {
      gtk_selection_data_set(selection_data, GDK_SELECTION_TYPE_STRING,
                            8, (unsigned char*) "", 0);
    }
  } else if (target == atomUtf8String || target == atomTextPlain || target == atomTextPlainUtf8) {
    if (contents->textPlain != nullptr) {
      gtk_selection_data_set_text(selection_data, contents->textPlain, strlen(contents->textPlain));
    } else {
      gtk_selection_data_set_text(selection_data, "", 0);
    }
  }
}


  // Setting clipboard and callback handler:

  GtkClipboard* clipboard =
      gtk_clipboard_get_default(gdk_display_get_default());
  if (clipboard == nullptr) {
    return FL_METHOD_RESPONSE(fl_method_error_response_new(
        kClipboardError, "Cannot get GTK clipboard", nullptr));
  }

  ClipboardContents *contents = (ClipboardContents *) malloc(sizeof(ClipboardContents));
  if (html_value != nullptr) {
    contents->textHtml = g_strdup(fl_value_get_string(html_value));
  }
  contents->textPlain = g_strdup(fl_value_get_string(text_value));

  GtkTargetList * list = gtk_target_list_new(NULL, 0);
  if (html_value != nullptr) {
    gtk_target_list_add(list, atomTextHtml, 0, 0);
  }
  gtk_target_list_add_text_targets(list, 0);
  int n_targets;
  GtkTargetEntry *targets = gtk_target_table_new_from_list(list, &n_targets);

  if (!gtk_clipboard_set_with_data(clipboard, targets, n_targets,
      (GtkClipboardGetFunc) clipboard_get_cb, (GtkClipboardClearFunc) clipboard_clear_cb,
      contents)) {
    gtk_target_table_free(targets, n_targets);
    gtk_target_list_unref(list);
    return FL_METHOD_RESPONSE(fl_method_error_response_new(
        kClipboardError, "Clipboard set with data failed", nullptr));
  }

  gtk_clipboard_set_can_store(clipboard, nullptr, 0);
  gtk_clipboard_store(clipboard);
	gtk_target_table_free(targets, n_targets);
	gtk_target_list_unref(list);
```
</details>

**Comment by chunhtai** [comment]
@tgucio 
If the purpose for html is to save multimedia, we could use the multimedia clipboard in each platform directly.
If we convert rich text to image, paste content will no longer be text. On the other hand, we could convert the rich text to html tag with css styling, but that is much more work than using the rich text clipboard.

**Comment by tgucio** [comment]
@chunhtai 
I guess the point I was trying to make was that HTML would be a future proof option allowing to also put images together with text on the clipboard (e.g. a TextSpan with plain text and WidgetSpans). For rich text, it would be pretty easy by wrapping plain text in `<span style...>`, e.g. `<span style="color:blue;font-size:24px">blue</span>` .

**Comment by chunhtai** [comment]
@tgucio That is interesting, I guess that will be great if we want to paste it into other app. This also has to work the other way around though, we may need to be able to parse pasted html back to the format Flutter can understand. I also need to verify whether it is possible to store html into clipboard and get them back in rich text or vise versa for all platform.

**Comment by tgucio** [comment]
@chunhtai
Sure, the main drawback is having to parse HTML although dart:html should do the job.
The snippets I pasted show HTML clipboard set methods, getting works in a similar fashion on each platform (verified copying/pasting back and forth in Flutter->Flutter, Flutter->apps and apps->Flutter scenarios with Chrome, Word and a few other apps).

**Comment by chunhtai** [comment]
pr merged, there are a bunch of other feature that are still missing, for example, rich text clipboard, keyboard arrow shortcuts, browser native context... and etc. I will create separate issue and label them under https://github.com/flutter/flutter/labels/f%3A%20selection. Closing this issue as done

**Comment by cnmade** [comment]
The fastest select text is smearing or draw to select text. 

For example. 

user can smearing on text area , to select text they want. want they smearing, is what they need. 

On mobile phone, IOS or Android,   using finger to smearing is faster than other solution.

![image](https://user-images.githubusercontent.com/278153/171529243-79bc1de6-3d71-48ce-97f4-ac4d2aafc934.png)
